refactor openresty locations and templates loading...

xmonader opened the pull request:
also use relative redirects and set a new variable called
where it contains  or  (to preserve port inside hostnames)

zaibon commented on Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py:
```diff
@@ -76,6 +144,8 @@ def reserve_s3(self, email, threebot_id, location, size=1, source=None, refund=N
                     - 1 => 500 GiB of storage
                     - 2 => 1000 GiB of storage
         :type size: int, optional
+        :param duration: number of months the reservation should be valid for
```
I wonder if the minimal granularity of a reservation should be a month?
@andhartl @despiegk any comment on that ?

zaibon commented on Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py:
```diff
@@ -63,7 +63,75 @@ def _grid_broker_pub_key(self):
+
+        return data.threebot_id, data_dict
+
+    def reservations_transactions_list(self):
```
I think this can be improved.
With the current code, every time you want to list the reservation made, we have to hit both the explorers and the notary.
We could improve that by saving the transaction ids of the reservation in bcdb.

So instead of walking on all the transaction every time, we could have something like:
- adapt the schema of the object and add a transaction id list
- reservations_transactions_list first check if the transaction id list is empty
- if empty do what the current code does, but also save the transaction id
- if not, just return the transactions from the list

With this logic, after we release this version, the user will only do the full walk of their transaction once. Then all the next time he will only get the id from bcdb which saves a lot of calls on the notary.

zaibon commented on Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py:
```diff
@@ -63,7 +63,75 @@ def _grid_broker_pub_key(self):
+                    raise e
+        return transactions
+
+    def reservation_extend(self, transaction, duration=1, source=None, refund=None):
```
This function also needs to check that the reservation is not already expired. We cannot extend a expired transaction because we cannot ensure the capacity has not been already sell to someone else in the mean time.

Which also mean we need to be a bit smart in how we compute the final expiration date because if someone extend a 1 month reservation for another month, but we are only 2 week in the first reservation, the final expiration data is in 1month and 2 weeks.

serboctor replied:
As discussed, I added this logic in the grid_broker template.

andhartl commented:
Would keep the one month granularity to start with.  


> On 26. Jun 2019, at 09:17, Christophe de Carvalho <notifications@github.com> wrote:
> 
> @zaibon requested changes on this pull request.
> 
> In Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py <https://github.com/threefoldtech/jumpscaleX/pull/649#discussion_r297510340>:
> 
> > @@ -76,6 +144,8 @@ def reserve_s3(self, email, threebot_id, location, size=1, source=None, refund=N
>                      - 1 => 500 GiB of storage
>                      - 2 => 1000 GiB of storage
>          :type size: int, optional
> +        :param duration: number of months the reservation should be valid for
> I wonder if the minimal granularity of a reservation should be a month?
> @andhartl <https://github.com/andhartl> @despiegk <https://github.com/despiegk> any comment on that ?
> 
> In Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py <https://github.com/threefoldtech/jumpscaleX/pull/649#discussion_r297513773>:
> 
> > @@ -63,7 +63,75 @@ def _grid_broker_pub_key(self):
>              self._grid_broker_pub_key_ = vk.to_curve25519_public_key()
>          return self._grid_broker_pub_key_
>  
> -    def reserve_s3(self, email, threebot_id, location, size=1, source=None, refund=None):
> +    def _notary_data_get(self, hash_key):
> +        data = self._notary_client.get(hash_key)
> +        sk = self._threebot_singing_key(data.threebot_id)
> +        pk = _signing_key_to_private_key(sk)
> +        box = nacl.public.Box(pk, self._grid_broker_pub_key)
> +        data_dict = j.data.serializers.msgpack.loads(box.decrypt(data.content))
> +
> +        return data.threebot_id, data_dict
> +
> +    def reservations_transactions_list(self):
> I think this can be improved.
> With the current code, every time you want to list the reservation made, we have to hit both the explorers and the notary.
> We could improve that by saving the transaction ids of the reservation in bcdb.
> 
> So instead of walking on all the transaction every time, we could have something like:
> 
> adapt the schema of the object and add a transaction id list
> reservations_transactions_list first check if the transaction id list is empty
> if empty do what the current code does, but also save the transaction id
> if not, just return the transactions from the list
> With this logic, after we release this version, the user will only do the full walk of their transaction once. Then all the next time he will only get the id from bcdb which saves a lot of calls on the notary.
> 
> In Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py <https://github.com/threefoldtech/jumpscaleX/pull/649#discussion_r297515215>:
> 
> > +        list all transactions that happened due to a reservation
> +        """
> +        transactions = list()
> +        for transaction in self._wallet.transactions:
> +            if not transaction.data:
> +                continue
> +            try:
> +                _, data_dict = self._notary_data_get(transaction.data.value.decode())
> +                if data_dict.get("type") in ["vm", "reverse_proxy", "namespace", "s3"]:
> +                    transactions.append(transaction)
> +            except Exception as e:
> +                if "reservation not found" not in str(e):
> +                    raise e
> +        return transactions
> +
> +    def reservation_extend(self, transaction, duration=1, source=None, refund=None):
> This function also needs to check that the reservation is not already expired. We cannot extend a expired transaction because we cannot ensure the capacity has not been already sell to someone else in the mean time.
> 
> Which also mean we need to be a bit smart in how we compute the final expiration date because if someone extend a 1 month reservation for another month, but we are only 2 week in the first reservation, the final expiration data is in 1month and 2 weeks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/threefoldtech/jumpscaleX/pull/649?email_source=notifications&email_token=AEN22KCX5GUSOI5W5Z4IDDTP4MJXDA5CNFSM4H3F7KGKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4VA4KY#pullrequestreview-254414379>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEN22KH5PRPR43NIJPAED5LP4MJXDANCNFSM4H3F7KGA>.
>

zaibon commented on Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py:
```diff
@@ -136,6 +213,8 @@ def reserve_zdb_namespace(
         :type threebot_id: string
         :param size: size of the namespace in GB, defaults to 1
         :type size: int, optional
+        :param duration: number of duration the reservation should be valid for
```
```suggestion
        :param duration: number of month the reservation should be valid for
```

serboctor commented:
@zaibon  tests done.

abom committed:
refactor openresty locations and templates loading...

also use relative redirects and set a new variable called
where it contains  or  (to preserve port inside hostnames)
